sidechain/consensus/slots/src/slot_stream.rs

```diff
@@ -0,0 +1,113 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+		http://www.apache.org/licenses/LICENSE-2.0
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+//! Slots functionality for Substrate.
+//!
+//! Some consensus algorithms have a concept of *slots*, which are intervals in
+//! time during which certain events can and/or must occur.  This crate
+//! provides generic functionality for slots.
+
+use crate::time_until_next_slot;
+use futures_timer::Delay;
+use std::time::Duration;
+
+/// Executes given `task` repeatedly when the next slot becomes available.
+pub async fn start_slot_worker<F>(task: F, slot_duration: Duration)
+where
+	F: Fn(),
+{
+	let mut slot_stream = SlotStream::new(slot_duration);
+
+	loop {
+		slot_stream.next_slot().await;
+		task();
+	}
+}
+
+/// Stream to calculate the slot schedule with.
+pub struct SlotStream {
+	slot_duration: Duration,
+	inner_delay: Option<Delay>,
+}
+
+impl SlotStream {
+	pub fn new(slot_duration: Duration) -> Self {
+		SlotStream { slot_duration, inner_delay: None }
+	}
+}
+
+/// Waits for the duration of `inner_delay`.
```

Comment:
Oh sorry, missed that one - though I won't loose any sleep if that isn't fixed : I think that should be a function description, not a struct impl one.

I.e. Right on top of `pub async fn next_slot(&mut self)`